reactioncommerce/reaction
# feat: decouple SMTP email sending logic from core to allow plugins to override

Resolves #4730 
Impact: **minor**  
Type: **feature**

## Issue
Currently, SMTP email sending logic is hard-coded in core. This makes it impossible to integrate with a third-party email provider that does not support SMTP, without modifying core code.

## Solution
This PR decouples SMTP from core by creating a new `reaction-email-smtp` included plugin. Core now emits a `sendEmail` event for each email job, which any plugin can respond to. The included SMTP plugin responds to it, checks if an SMTP email provider is configured, and sends the email if so.

The email provider config form found at Dashboard -> Emails -> Mail Provider is now also able to be overridden. Plugins can use register.js to provide a React component to use here.


## Breaking changes
None


## Testing
1. Login as operator -> Dashboard -> Email -> Mail Provider. Confirm the existing config component works as expected (shows status icon & existing settings, gear icon shows config form, you can successfully configure an SMTP provider, etc)
2. After configuring an SMTP provider, complete an order and confirm the email was successfully delivered
3. Confirm in database that Emails.status is set to "completed"
4. Set SMTP settings to something invalid & save
5. Complete another order, confirm email error in logs, and that Emails.status is set to "failed" in database
6. Create an `email-provider-test` plugin in `imports/plugins/custom` with 3 files:
- register.js
```
import Reaction from "/imports/plugins/core/core/server/Reaction";

Reaction.registerPackage({
  label: "Email Provider Test",
  name: "reaction-email-provider-test",
  icon: "fa fa-envelope-o",
  autoEnable: true,
  registry: [
    {
      provides: ["emailProviderConfig"],
      template: "TestEmailConfig"
    }
  ]
});
```
- client/TestEmailConfig.js:
```
import React, { Component } from "react";
import { registerComponent } from "@reactioncommerce/reaction-components";

class TestEmailConfig extends Component {
  render() {
    return (
      <div>Test email provider config here</div>
    );
  }
}

registerComponent("TestEmailConfig", TestEmailConfig);
```
- client/index.js:
```
import "./TestEmailConfig";
```
7. Restart Reaction and confirm you see "Test email provider config here" at Dashboard -> Email -> Mail Provider
